File: 04-Prime-Sieve-Factorisation/02-Segmented-Sieve/java/SegmentedSieve.java
/*
 * Problem: Segmented Sieve - Primes in Range [L, R]
 * Source:  SPOJ PRIME1
 * Tags: segmented-sieve, prime
 */

import java.util.*;

public class SegmentedSieve {
    public static List<Long> segmentedSieve(long L, long R) {
        long sqrtR = (long) Math.sqrt(R) + 1;
        boolean[] smallSieve = new boolean[(int) sqrtR + 1];
        Arrays.fill(smallSieve, true);
        smallSieve[0] = smallSieve[1] = false;
        for (long i = 2; i * i <= sqrtR; i++)
            if (smallSieve[(int) i])
                for (long j = i * i; j <= sqrtR; j += i)
                    smallSieve[(int) j] = false;
        
        boolean[] isPrime = new boolean[(int)(R - L + 1)];
        Arrays.fill(isPrime, true);
        if (L <= 1) isPrime[(int)(1 - L)] = false;
        
        for (long i = 2; i <= sqrtR; i++) {
            if (!smallSieve[(int) i]) continue;
            long start = Math.max(i * i, ((L + i - 1) / i) * i);
            for (long j = start; j <= R; j += i)
                isPrime[(int)(j - L)] = false;
        }
        
        List<Long> primes = new ArrayList<>();
        for (long i = Math.max(2L, L); i <= R; i++)
            if (isPrime[(int)(i - L)]) primes.add(i);
        return primes;
    }
    
    public static void main(String[] args) {
        System.out.println("Primes in [10,50]: " + segmentedSieve(10, 50));
    }
}

File: 04-Prime-Sieve-Factorisation/02-Segmented-Sieve/cpp/segmented_sieve.cpp
/*
 * Problem: Segmented Sieve - Primes in Range [L, R]
 * Source:  SPOJ PRIME1 - https://www.spoj.com/problems/PRIME1/
 * 
 * Approach: Sieve small primes up to √R, then use them to mark composites in [L, R].
 * Time Complexity:  O((R-L+1) * log(log R) + √R * log(log √R))
 * Tags: segmented-sieve, prime
 */

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

vector<ll> segmentedSieve(ll L, ll R) {
    ll sqrtR = (ll)sqrt((double)R) + 1;
    
    // Step 1: Get small primes up to √R
    vector<bool> smallSieve(sqrtR + 1, true);
    smallSieve[0] = smallSieve[1] = false;
    for (ll i = 2; i * i <= sqrtR; i++)
        if (smallSieve[i])
            for (ll j = i * i; j <= sqrtR; j += i)
                smallSieve[j] = false;
    
    // Step 2: Mark composites in [L, R]
    vector<bool> isPrime(R - L + 1, true);
    if (L <= 1) isPrime[1 - L] = false; // 1 is not prime
    
    for (ll i = 2; i <= sqrtR; i++) {
        if (!smallSieve[i]) continue;
        ll start = max(i * i, ((L + i - 1) / i) * i);
        for (ll j = start; j <= R; j += i)
            isPrime[j - L] = false;
    }
    
    vector<ll> primes;
    for (ll i = max(2LL, L); i <= R; i++)
        if (isPrime[i - L])
            primes.push_back(i);
    
    return primes;
}

int main() {
    auto primes = segmentedSieve(10, 50);
    cout << "Primes in [10, 50]: ";
    for (ll p : primes) cout << p << " ";
    cout << endl;
    // Output: 11 13 17 19 23 29 31 37 41 43 47
    return 0;
}
